File: solutions/_554_brick_wall/solution.py
from typing import List


class Solution:
    def leastBricks(self, wall: List[List[int]]) -> int:
        idx2gaps = {0: 0}
        max_gaps = 0

        for r in wall:
            idx = 0
            for b in r[:-1]:
                idx += b
                idx2gaps[idx] = 1 + idx2gaps.get(idx, 0)
                max_gaps = max(max_gaps, idx2gaps[idx])

        return len(wall) - max_gaps


# https://leetcode.com/problems/brick-wall/

File: solutions/_554_brick_wall/solution.java
package _554_brick_wall;

import java.util.*;

class Solution {

  public int leastBricks(List<List<Integer>> wall) {
    Map<Integer, Integer> idx2gaps = new HashMap<>();
    int max = 0;

    for (List<Integer> row : wall) {
      int idx = 0;
      for (int i = 0; i < row.size() - 1; i++) {
        idx += row.get(i);
        idx2gaps.put(idx, idx2gaps.getOrDefault(idx, 0) + 1);
        max = Math.max(max, idx2gaps.get(idx));
      }
    }

    return wall.size() - max;
  }
}
/*
https://leetcode.com/problems/brick-wall/
*/
